src/MOI_wrapper.jl

```diff
@@ -472,24 +448,93 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
+    π(v::Vector{Float64}, model::MOI.ModelLike, conic_form::MatOI.GeometricConicForm, index_map::MOIU.IndexMap)
 
-Find projection of vectors in `v` on product of `cones`.
-For more info, refer https://github.com/matbesancon/MathOptSetDistances.jl
+Given a `model`, its `conic_form` and the `index_map` from the indices of
+`model` to the indices of `conic_form`, find the projection of the vectors `v`
+of length equal to the number of rows in the conic form onto the cartesian
+product of the cones corresponding to these rows.
+For more info, refer to https://github.com/matbesancon/MathOptSetDistances.jl
 """
-π(cones, v) = MOSD.projection_on_set(MOSD.DefaultDistance(), v, cones)
+function π(v::Vector{T}, model::MOI.ModelLike, conic_form::MatOI.GeometricConicForm, index_map::MOIU.IndexMap) where T
+    return map_rows(model, conic_form, index_map, Flattened{T}()) do ci, r
+        MOSD.projection_on_set(
+            MOSD.DefaultDistance(),
+            v[r],
+            MOI.dual_set(MOI.get(model, MOI.ConstraintSet(), ci))
+        )
+    end
+end
 
 
 """
-    Dπ(cones::Array{<: MOI.AbstractVectorSet}, v::Vector{Vector{Float64}})
+    Dπ(v::Vector{Float64}, model, conic_form::MatOI.GeometricConicForm, index_map::MOIU.IndexMap)
 
-Find gradient of projection of vectors in `v` on product of `cones`.
-For more info, refer https://github.com/matbesancon/MathOptSetDistances.jl
+Given a `model`, its `conic_form` and the `index_map` from the indices of
+`model` to the indices of `conic_form`, find the gradient of the projection of
+the vectors `v` of length equal to the number of rows in the conic form onto the
+cartesian product of the cones corresponding to these rows.
+For more info, refer to https://github.com/matbesancon/MathOptSetDistances.jl
 """
-Dπ(cones, v) = MOSD.projection_gradient_on_set(MOSD.DefaultDistance(), v, cones)
+function Dπ(v::Vector{T}, model::MOI.ModelLike, conic_form::MatOI.GeometricConicForm, index_map::MOIU.IndexMap) where T
+    return BlockDiagonals.BlockDiagonal(
+        map_rows(model, conic_form, index_map, Nested{Matrix{T}}()) do ci, r
+            MOSD.projection_gradient_on_set(
+                MOSD.DefaultDistance(),
+                v[r],
+                MOI.dual_set(MOI.get(model, MOI.ConstraintSet(), ci)),
+            )
+        end
+    )
+end
+
+struct Nested{T} end
+struct Flattened{T} end
+
+function _assign_mapped!(x, y, r, k, ::Nested)
+    x[k] = y
+end
+function _assign_mapped!(x, y, r, k, ::Flattened)
+    x[r] = y
+end
+
+function _map_rows!(f::Function, x::Vector, model, conic_form::MatOI.GeometricConicForm, index_map::MOIU.DoubleDicts.IndexWithType{F, S}, map_mode, k) where {F, S}
+    for ci in MOI.get(model, MOI.ListOfConstraintIndices{F, S}())
+        r = MatOI.rows(conic_form, index_map[ci])
+        k += 1
+        _assign_mapped!(x, f(ci, r), r, k, map_mode)
+    end
+    return k
+end
+
+_map_output(conic_form, ::Nested{T}) where {T} = Vector{T}(undef, length(conic_form.dimension))
```

Comment:
this function name does not reflect what it does, "initialize_x" would be better? Not sure what x could be here

Reply:
Done in latest commit